driver-reactive-streams/src/main/com/mongodb/reactivestreams/client/internal/BatchCursorFlux.java

```diff
@@ -0,0 +1,120 @@
+
+    private final BatchCursorPublisher<T> batchCursorPublisher;
+    private final AtomicBoolean inProgress = new AtomicBoolean(false);
+    private final AtomicLong demandDelta = new AtomicLong(0);
+    private BatchCursor<T> batchCursor;
+    private FluxSink<T> sink;
+
+    BatchCursorFlux(final BatchCursorPublisher<T> batchCursorPublisher) {
+        this.batchCursorPublisher = batchCursorPublisher;
+    }
+
+    @Override
+    public void subscribe(final Subscriber<? super T> subscriber) {
+        Flux.<T>create(sink -> {
+            this.sink = sink;
+            sink.onRequest(demand -> {
+                if (calculateDemand(demand) > 0 && inProgress.compareAndSet(false, true)) {
+                    if (batchCursor == null) {
+                        int batchSize = calculateBatchSize(sink.requestedFromDownstream());
+                        batchCursorPublisher.batchCursor(batchSize).subscribe(bc -> {
+                            batchCursor = bc;
+                            inProgress.set(false);
+                            recurseCursor();
+                        }, sink::error);
+                    } else {
+                        inProgress.set(false);
+                        recurseCursor();
+                    }
+                }
+            });
+            sink.onCancel(this::closeCursor);
+            sink.onDispose(this::closeCursor);
+        }, FluxSink.OverflowStrategy.BUFFER)
+        .subscribe(subscriber);
+    }
+
+    private void closeCursor() {
+        if (batchCursor != null && !batchCursor.isClosed()) {
+            batchCursor.close();
+        }
+    }
+
+    private void recurseCursor(){
+        if (!sink.isCancelled() && sink.requestedFromDownstream() > 0 && inProgress.compareAndSet(false, true)) {
+            if (batchCursor.isClosed()) {
+                sink.complete();
+            } else {
+                batchCursor.setBatchSize(calculateBatchSize(sink.requestedFromDownstream()));
+                Mono.from(batchCursor.next())
+                        .doOnCancel(this::closeCursor)
+                        .doOnError((e) -> {
+                            closeCursor();
+                            sink.error(e);
+                        })
+                        .doOnSuccess(results -> {
+                            if (results != null) {
+                                results.forEach(sink::next);
+                                calculateDemand(-results.size());
+                            }
+                            if (batchCursor.isClosed()) {
+                                sink.complete();
+                            } else {
+                                inProgress.set(false);
+                                recurseCursor();
+                            }
+                        })
+                        .subscribe();
+                }
+        }
+    }
+
+    synchronized long calculateDemand(final long demand) {
```

Comment:
I propose using `AtomicLong.accumulateAndGet` method, which is meant to solve tasks like this one.
```java
long calculateDemand(final long demand) {
    return demandDelta.accumulateAndGet(demand, (originalDelta, update) -> {
        long newDelta = originalDelta + update;
        return demand > 0 && newDelta < originalDelta ? Long.MAX_VALUE : newDelta;
    });
}
```

Reply:
Nice - Done